### ISSUE TITLE
laravel cors is not working with laravel 5.5

### ISSUE THREAD

**Issue by swapnil-chaudhari** [fruitcake/laravel-cors#278]
Hi,
I am creating demo app with React Js as frontend and Laravel 5.5 as backend. I have followed all the steps you mentioned for cors. I am able to get data properly with GET request but,  other requests like POST is not working.

```
config/cors.php

    'supportsCredentials' => false,
    'allowedOrigins' => ['*'],
    'allowedOriginsPatterns' => [],
    'allowedHeaders' => ['*'],
    'allowedMethods' => ['*'],
    'exposedHeaders' => [],
    'maxAge' => 0,
```

```
axios call from react

export const headers = {
    headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
    }
};

const sendCategory = category => {
    if (!category.categoryId) {
        console.log(category);
        return axios.post(CATEGORY_ENDPOINT, category, headers);
    }

    return axios.put(CATEGORY_ENDPOINT + `?id=${category.categoryId}`, category, headers)
}
```

![image](https://user-images.githubusercontent.com/17541314/34598570-9b2421da-f213-11e7-8406-fd57bc021965.png)

![image](https://user-images.githubusercontent.com/17541314/34598582-ab2fd704-f213-11e7-8b5c-42520d2f9fc3.png)

![image](https://user-images.githubusercontent.com/17541314/34598590-b309c5d4-f213-11e7-80db-f40bfc8bb7c1.png)

is there anything i am missing here? If it works for GET request then it should work for POST as well.

**Comment by barryvdh** [comment]
Did you disable CSRF protection?

**Comment by swapnil-chaudhari-1811** [comment]
you mean removing of `\App\Http\Middleware\VerifyCsrfToken::class,` from kernel.php?

**Comment by swapnil-chaudhari-1811** [comment]
@barryvdh Please help me in this, what is missing in this?

**Comment by cpuzzuol** [comment]
I am getting the 500 error "Class cors does not exist" after upgrading from 5.4 to 5.5

**Comment by cpuzzuol** [comment]
so if you're calling the middleware 'cors' in your controller, you need to add it to you $middlewareGroups in Kernel.php:

'cors' => [
           \Barryvdh\Cors\HandleCors::class,
        ],

This solved the problem at least in my case.

**Comment by acacha** [comment]
@barryvdh same problem here

**Comment by acacha** [comment]
Solved but I'm not sure how but I think maybe the problem is that a previos laravel app was executed in same domain and port (localhost:8080) an now my vue-cli app is running in the same domain (localhost:8080) but my broser remember old Laravel session and cookies so is sending a CSRF token?

**Comment by acacha** [comment]
About my last comment I try to reproduce the problem but it not seems the problem because all works ok... Sorry I'm still don't know what I have changed to solve my problem...

**Comment by acacha** [comment]
My working kernel.php config:

```
<?php

namespace App\Http;

use Illuminate\Foundation\Http\Kernel as HttpKernel;

class Kernel extends HttpKernel
{
    /**
     * The application's global HTTP middleware stack.
     *
     * These middleware are run during every request to your application.
     *
     * @var array
     */
    protected $middleware = [
        \Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
        \App\Http\Middleware\TrustProxies::class,
        \Barryvdh\Cors\HandleCors::class,
    ];

    /**
     * The application's route middleware groups.
     *
     * @var array
     */
    protected $middlewareGroups = [
        'web' => [
            \App\Http\Middleware\EncryptCookies::class,
            \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
            \Illuminate\Session\Middleware\StartSession::class,
            // \Illuminate\Session\Middleware\AuthenticateSession::class,
            \Illuminate\View\Middleware\ShareErrorsFromSession::class,
            \App\Http\Middleware\VerifyCsrfToken::class,
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
            \Laravel\Passport\Http\Middleware\CreateFreshApiToken::class,
        ],

        'api' => [
            'throttle:60,1',
            'bindings'
        ],
    ];

    /**
     * The application's route middleware.
     *
     * These middleware may be assigned to groups or used individually.
     *
     * @var array
     */
    protected $routeMiddleware = [
        'auth'       => \Illuminate\Auth\Middleware\Authenticate::class,
        'auth.basic' => \Illuminate\Auth\Middleware\AuthenticateWithBasicAuth::class,
        'bindings'   => \Illuminate\Routing\Middleware\SubstituteBindings::class,
        'can'        => \Illuminate\Auth\Middleware\Authorize::class,
        'guest'      => \App\Http\Middleware\RedirectIfAuthenticated::class,
        'throttle'   => \Illuminate\Routing\Middleware\ThrottleRequests::class,
    ];
}
```

**Comment by ashgibson** [comment]
This might help. I just upgraded to latest version and had an issue. I am setting my allowedOrigins using the following to pull values from .env.

`'allowedOrigins' => explode(',', env('ALLOWED_ORIGINS', ''))`

My env file has the following:

`ALLOWED_ORIGINS="localhost:4200,localhost:3000"`

But I've had to update it to the following to get it to work with the latest version of this library:

`ALLOWED_ORIGINS="http://localhost:4200,http://localhost:3000"`

**Comment by acacha** [comment]
Ok I've found the problem (in my case). As seen in docs:

https://github.com/barryvdh/laravel-cors#common-problems-and-errors-pre-laravel-53

"When an error occurs, the middleware isn't run completely."

So any error could returns a CORS problem because middleware CORS doesn't runs oks. This but don't mean anything is wrong with CORS it could be any other problem.

In my case CORS doesn't works because I'm dumping (for debugging reasons) values to api output so I provoke a Headers already sent error.

**Comment by SajjadKharrazi** [comment]
same problem

**Comment by SajjadKharrazi** [comment]
anyone????

**Comment by CrixuAMG** [comment]
Experiencing the same issue using a (freshly made) Laravel 5.5 API and a VueJS dashboard.

**Comment by John-Ciuchea** [comment]
Had same issue. solved by registering the middleware in kernel. odd thing is that if your api has a error or you have a dd in your code the Access-Control-Allow-Origin error is there, this might be an axios issue.

**Comment by CrixuAMG** [comment]
@John-Ciuchea in which array did you register it?

**Comment by John-Ciuchea** [comment]
@CrixuAMG I register the cors middleware only for api group but you can add it in the global.

**Comment by malikali86** [comment]
Had the same issue, tried every thing related to changing code, configurations or cors settings, nothing worked. Just changed my PHP version from 7.0.23 to 7.1.9 and it works!

In simple words, could be to do something with your PHP settings or version, try changing it to some other version to see if it works. As for settings, I followed standard installation instructions on the githum plugin page.

**Comment by malikali86** [comment]
I had the same issue, but when I added the code, shown below, at the top of the Laravel routes file (routes\api.php) it started working again.

header('Access-Control-Allow-Credentials: true');

Best of luck.

**Comment by UT-Tom** [comment]
nothing works yayyy

![](https://cdn.thomasj.me/assets/img/500.gif)

**Comment by malikali86** [comment]
Well another temporary fix which worked for me is to restart the wamp server on my local wamp environment, or change PHP version. But it is so random, I have to repeat this trick every day or two.

I have no idea what is the real core issue, hoping it will get resolved by the plugin developers by the time I finish my coding. In the meantime the trick above works for me.

**Comment by bb128b** [comment]
@John-Ciuchea this solution worked for me rather than putting the class name directly in the 'api' group.

**Comment by yoezasapoetro** [comment]
I've used this package and didnt works for laravel 5.5 so to solve the problem set header into .htaccess manually then it works
Ps: i use apache server on my machine

**Comment by CrixuAMG** [comment]
@yoezasapoetro since I'm running it on Laravel's Homestead vm a .htaccess is not available, it's also not available for nginx environments, only for Apache servers.

**Comment by yoezasapoetro** [comment]
i was found cors issue after i put my code on production machine, not homestead person @CrixuAMG

**Comment by CrixuAMG** [comment]
@yoezasapoetro just notifying people that the solution you provided might not work for all environments :)

**Comment by usmanzix** [comment]
I am also getting the same issue in laravel 5.5 
```
Failed to load http://127.0.0.1:8000/api/user: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access. The response had HTTP status code 403.
```

**Comment by ingria** [comment]
Confirming, no headers on Laravel 5.5:

```
> curl -I 127.0.0.1:8000
HTTP/1.1 200 OK
Host: 127.0.0.1:8000
Connection: close
X-Powered-By: PHP/7.0.25-0ubuntu0.16.04.1
Cache-Control: no-cache, private
Date: Thu, 22 Feb 2018 22:09:26 GMT
Content-Type: text/html; charset=UTF-8
```

**Comment by vjrngn** [comment]
Moving the registration of `HandleCors::class ` from `$routeMiddleware` to `$middleware` solved my problem and I was able to successfully intercept the 401 response on my web client. Hope this helps someone.

I'm using this cors package along with tymon-jwt and I think the problem was caused due to wrong order of execution of the middleware. 

*Previous Config*
```php
protected $routeMiddleware = [
    'auth'       => Authenticate::class,
    'auth.basic' => AuthenticateWithBasicAuth::class,
    'bindings'   => SubstituteBindings::class,
    'can'        => Authorize::class,
    'guest'      => RedirectIfAuthenticated::class,
    'throttle'   => ThrottleRequests::class,
    'cors'       => HandleCors::class,
];
```

*Fix*
```php
 protected $middleware = [
    CheckForMaintenanceMode::class,
    ValidatePostSize::class,
    TrimStrings::class,
    ConvertEmptyStringsToNull::class,
    TrustProxies::class,
    HandleCors::class
];

protected $routeMiddleware = [
    'auth'       => Authenticate::class,
    'auth.basic' => AuthenticateWithBasicAuth::class,
    'bindings'   => SubstituteBindings::class,
    'can'        => Authorize::class,
    'guest'      => RedirectIfAuthenticated::class,
    'throttle'   => ThrottleRequests::class,
];
```

**Comment by hrabbit** [comment]
I'm hugely appreciative of this solution which got me back up and running again.. to that, hats off everybody :tophat: 

However, I'm rather confused about why this is required to actually work to be honest. 

Specifically..

```
protected $middleware = [
    HandleCors::class
];
```

Basically it's just moving the header to be present on every single request (api and/or web, as well as every single request path), which in some cases might be fine, but other times sounds like a bad thing (ex. I have many different routes/domains configured with different restrictions) that require different cors rules applied. I realise there are ways around that, just an observation.

I haven't dug into the cause myself in my own instances but I'd be interested to know if any of the developers/contributes have seen/intercepted this issue yet.

**Comment by Tarasovych** [comment]
I wish @hrabbit last commend will be added to [configuration](https://github.com/barryvdh/laravel-cors#configuration)

**Comment by andyhernandez** [comment]
@hrabbit I had the same exact issue as you.  The plugin only works when applied to the global HTTP middleware stack.  I'm using this as a temporary fix, but I definitely do not want to go to production with the entire API being exposed to CORS.

**Comment by kyawthet-flc** [comment]
> 
> 
> Ok I've found the problem (in my case). As seen in docs:
> 
> https://github.com/barryvdh/laravel-cors#common-problems-and-errors-pre-laravel-53
> 
> "When an error occurs, the middleware isn't run completely."
> 
> So any error could returns a CORS problem because middleware CORS doesn't runs oks. This but don't mean anything is wrong with CORS it could be any other problem.
> 
> In my case CORS doesn't works because I'm dumping (for debugging reasons) values to api output so I provoke a Headers already sent error.

I was also like you did.It took me 9 hrs!

**Comment by fts1025** [comment]
Im my case, the problem was solved if the order of API middleware is as follows.

```php
protected $middlewareGroups = [
    'api' => [
        \Barryvdh\Cors\HandleCors::class,
        'throttle:60,1',
        'bindings',
    ],
];
```

**Comment by anbu369** [comment]
> Moving the registration of `HandleCors::class ` from `$routeMiddleware` to `$middleware` solved my problem and I was able to successfully intercept the 401 response on my web client. Hope this helps someone.
> 
> I'm using this cors package along with tymon-jwt and I think the problem was caused due to wrong order of execution of the middleware.
> 
> _Previous Config_
> 
> ```
> protected $routeMiddleware = [
>     'auth'       => Authenticate::class,
>     'auth.basic' => AuthenticateWithBasicAuth::class,
>     'bindings'   => SubstituteBindings::class,
>     'can'        => Authorize::class,
>     'guest'      => RedirectIfAuthenticated::class,
>     'throttle'   => ThrottleRequests::class,
>     'cors'       => HandleCors::class,
> ];
> ```
> 
> _Fix_
> 
> ```
>  protected $middleware = [
>     CheckForMaintenanceMode::class,
>     ValidatePostSize::class,
>     TrimStrings::class,
>     ConvertEmptyStringsToNull::class,
>     TrustProxies::class,
>     HandleCors::class
> ];
> 
> protected $routeMiddleware = [
>     'auth'       => Authenticate::class,
>     'auth.basic' => AuthenticateWithBasicAuth::class,
>     'bindings'   => SubstituteBindings::class,
>     'can'        => Authorize::class,
>     'guest'      => RedirectIfAuthenticated::class,
>     'throttle'   => ThrottleRequests::class,
> ];
> ```

I got the error after using jwt and this fixed it. thanks

**Comment by John-Ciuchea** [comment]
Registering the middleware on `$middleware` array will enable CORS for all routes (api, web ...). 
If your clients are calling only api endpoints then I don't think this should be enabled globally.

**Comment by josegus** [comment]
same here, nothing is working, i've tried everything

**Comment by hrabbit** [comment]
> Im my case, the problem was solved if the order of API middleware is as follows.
> 
> ```
> protected $middlewareGroups = [
>     'api' => [
>         \Barryvdh\Cors\HandleCors::class,
>         'throttle:60,1',
>         'bindings',
>     ],
> ];
> ```

I believe this is my own method of performing cors also, however, I register the class within 
```
protected $routeMiddleware = [
   ...
   'cors' => \Barryvdh\Cors\HandleCors::class,
];
```

Then just enable it inside the middlewareGroups
```
protected $middlewareGroups = [
  'web' => [
    ...
  ],
  'api' => [
    'cors',
    'throttle:60,1',
    'bindings',
  ],
];
```

**Comment by AndrewSavetchuk** [comment]
Same here...

**Comment by Rainesdave** [comment]
Same here, Laravel 6 and not working

**Comment by YaangVu** [comment]
> Same here, Laravel 6 and not working

yep, me too

**Comment by josegus** [comment]
any idea how can we fix this, @barryvdh ?

**Comment by YaangVu** [comment]
> any idea how can we fix this, @barryvdh ?

I add  `\Barryvdh\Cors\HandleCors::class,`    into `$middleware` array, and it work :D

**Comment by josegus** [comment]
> > any idea how can we fix this, @barryvdh ?
> 
> I add `\Barryvdh\Cors\HandleCors::class,` into `$middleware` array, and it work :D

yes @YaangVu, we all did the same. What we are trying to figure out is why is it not working in certain situations (may be a conflict with another middleware or something else). In my case, I'm working with laravel 5.8 and this is the only extra middleware I'm using

**Comment by nmukumba** [comment]
For me I saw that I was getting CORS because I was dumping data for debug purposes using the die and dump method.

**Comment by YaangVu** [comment]
> > > any idea how can we fix this, @barryvdh ?
> > 
> > 
> > I add `\Barryvdh\Cors\HandleCors::class,` into `$middleware` array, and it work :D
> 
> yes @YaangVu, we all did the same. What we are trying to figure out is why is it not working in certain situations (may be a conflict with another middleware or something else). In my case, I'm working with laravel 5.8 and this is the only extra middleware I'm using

oh sorry, I'm wrong. It still not working :'(

**Comment by YaangVu** [comment]
I use Nginx as Web server. And add cors both of `nginx config` and `laravel cors filter`. And It not work.
Moment, I remove cors in `nginx config` and my Application can work normally.

@josegus

**Comment by 949mac** [comment]
> For me I saw that I was getting CORS because I was dumping data for debug purposes using the die and dump method.

I had this same issue. The `dump()` command has foiled me before. Thanks for the suggestion here. After removing my dump statements, CORS worked again, right away.

**Comment by kburudi** [comment]
I have been having this issue on a new API and react front-end am working on for the last one week. I am using Laravel 6. I had done a previous API about two months ago still on Laravel 6 and react and everything was ok. To keep the long story short, I realized the issue was with my frontend and am trying to figure out what exactly. I turned my backend server off and sent the request and still got the cors error 

```Access to XMLHttpRequest at 'localhost:8000/api/v1/auth/login' from origin 'http://localhost:3000' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.```

So I decided to test this same login with another frontend and it has worked. I know that this may not be the case for some of you but you should probably check it out. The backend as I have relized has been ok all along and its an issue with my react code.

**Comment by kburudi** [comment]
> I have been having this issue on a new API and react front-end am working on for the last one week. I am using Laravel 6. I had done a previous API about two months ago still on Laravel 6 and react and everything was ok. To keep the long story short, I realized the issue was with my frontend and am trying to figure out what exactly. I turned my backend server off and sent the request and still got the cors error
> 
> `Access to XMLHttpRequest at 'localhost:8000/api/v1/auth/login' from origin 'http://localhost:3000' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.`
> 
> So I decided to test this same login with another frontend and it has worked. I know that this may not be the case for some of you but you should probably check it out. The backend as I have realized has been ok all along and its an issue with my react code.

So I decided to bypass the cors using `https://cors-anywhere.herokuapp.com` and ran into this new error 
```Not found because of proxy error: Error: connect ECONNREFUSED 127.0.0.1:8000```
This error occurs because the frontend does hit my backend server. So I google and one solution states I add an absolute path to my request. Since I am using Axios, I had created an instance. 
![Screenshot from 2019-12-21 13-27-20](https://user-images.githubusercontent.com/45054246/71306840-a8d9e280-23f6-11ea-851d-5c9655a4278a.png)
![Screenshot from 2019-12-21 13-27-43](https://user-images.githubusercontent.com/45054246/71306844-ae372d00-23f6-11ea-95d7-542fb1b47f52.png)

axios.js
```
import axios from 'axios';
import promise from 'promise';

const instance = axios.create({
    baseURL: process.env.REACT_APP_BACKEND_URL
});

instance.interceptors.request.use((config) => {

    config.headers = {
      "content-type": "application/json",
      Accept: "application/json",
    };

    return config;
}, error => promise.reject(error));

export default instance;
```

I removed the instance and made my action like below.
action.js
```
import * as constants from "../Constants";
// import instance from "path_to_my_instance";  --- using my intance ---
import axios from "axios"; // using default axios

export default userData => dispatch => {
  console.log(axios.defaults);
  dispatch({
    type: constants.LOGIN_REQUEST
  });
    // instance.post("/auth/login", userData) --- Before (not working)---
   axios
    .post("http://localhost:8000/api/v1/auth/login", userData) // working
    .then(response => {
      dispatch({
        type: constants.LOGIN_SUCCESS,
        payload: response.data
      });
      console.log(response.data);
    })
    .catch(error => {
      dispatch({
        type: constants.LOGIN_ERROR,
        error
      });
    });
};
```
![Screenshot from 2019-12-21 13-40-06](https://user-images.githubusercontent.com/45054246/71306872-6fee3d80-23f7-11ea-8ff4-3cced4b2aa60.png)


That was my error and that's my solution this far. Glad if it helps any of you.

**Comment by zlanich** [comment]
I too am having this issue. I'm using https://github.com/specialtactics/laravel-api-boilerplate, so it's Dingo API and Laravel. The CORS middleware is not returning CORS headers, no matter where I put it in the middleware arrays. This is causing weird issues in axios in my React Native app, where 401 errors are coming back with the response object undefined.

**Comment by barryvdh** [comment]
Please use the latest version, follow the changes described in the readme and make sure to:
- use latest 1.x
- use global middleware (not route/group)
- set your paths in the config (new option)
- clear your config cache

Otherwise please create a new issue.
